src/Setup.ml

```diff
@@ -0,0 +1,215 @@
+           |> then_ (function
+                | Ok () -> Ok () |> resolve
+                | Error Fs.E.PathNotFound ->
+                  Error (E.InvalidPath hiddenEsyRoot) |> resolve))
+    |> then_ (function
+         | Error e -> Error e |> resolve
+         | Ok () ->
+           Filename.concat hiddenEsyRoot "esy.json"
+           |> dropAnEsyJSON
+           |> then_ (fun () -> resolve (Ok ())))
+    |> then_ (function
+         | Error e -> Error e |> resolve
+         | Ok () ->
+           Cmd.output esyCmd ~cwd:hiddenEsyRoot
+             ~args:[| "install"; "-P"; hiddenEsyRoot |]
+           |> then_ (function
+                | Error msg -> Error (E.EsyInstallFailure msg) |> resolve
+                | Ok _stdout ->
+                  reportProgress eventEmitter 0.1;
+                  AzurePipelines.getBuildID ()
+                  |> then_ (function
+                       | Ok id ->
+                         id |> AzurePipelines.getDownloadURL
+                         |> Js.Promise.then_ (function
+                              | Ok url -> resolve (Ok url)
+                              | Error msg ->
+                                Error (E.CacheFailure msg) |> resolve)
+                       | Error msg -> Error (E.CacheFailure msg) |> resolve)
+                  |> then_ (function
+                       | Ok url -> Ok url |> resolve
+                       | Error e -> Error e |> resolve)))
+    |> then_ (function
+         | Ok downloadUrl ->
+           Js.log2 "download" downloadUrl;
+           let lastProgress = ref 0.0 in
+           Js.Promise.make (fun ~resolve ~reject:_ ->
+               download downloadUrl
+                 (Path.join [| hiddenEsyRoot; "cache.zip" |])
+                 ~progress:(fun progressFraction ->
+                   let percent = progressFraction *. 80.0 in
+                   reportProgress eventEmitter (percent -. !lastProgress);
+                   lastProgress := percent)
+                 ~data:(fun _ -> ())
+                 ~error:(fun e ->
+                   (resolve (Error (E.CacheFailure e##message)) [@bs]))
+                 ~end_:(fun () -> (resolve (Ok ()) [@bs])))
+         | Error cacheFailureReason -> resolve (Error cacheFailureReason))
+    |> then_ (function
+         | Ok () ->
+           reportProgress eventEmitter 93.33;
+           Cmd.make ~cmd:"unzip" ~env:envWithUnzip
+           |> Js.Promise.then_ (function
+                | Error msg -> Js.Promise.resolve (Error msg)
+                | Ok unzip ->
+                  Cmd.output unzip ~args:[| "cache.zip" |] ~cwd:hiddenEsyRoot)
+           |> then_ (function
+                | Error _unzipCmdError ->
+                  Error (E.CacheFailure "Failed to unzip downloaded cache")
+                  |> resolve
+                | Ok _unzipCmdOutput -> resolve (Ok ()))
+         | Error e -> Error e |> resolve)
+    |> then_ (function
+         | Ok () ->
+           reportProgress eventEmitter 96.66;
```

Comment:
Where does this number come from?

Reply:
The progress reporting seen here is not perfect. I was hoping that Azure's API's would set the zip file's size in the headers, but it doesn't. This meant I had to hard code progress as a stopgap measure.

Reply:
Okay. Let's leave that as a comment then.